crates/extensions/c8y_http_proxy/src/actor.rs

```diff
@@ -162,80 +165,146 @@ impl C8YHttpProxyActor {
+        if self
+            .end_point
+            .get_internal_id(Some(device_id.as_str()))
+            .is_none()
+        {
+            let internal_id = self.try_get_internal_id(device_id.clone()).await?;
+            self.end_point
+                .set_internal_id(self.end_point.device_id.clone(), internal_id);
+        }
         Ok(())
     }
 
-    async fn get_c8y_internal_child_id(
-        &mut self,
-        child_device_id: String,
-    ) -> Result<String, C8YRestError> {
-        if let Some(c8y_internal_id) = self.child_devices.get(&child_device_id) {
-            Ok(c8y_internal_id.clone())
-        } else {
-            let c8y_internal_id = self.try_get_internal_id(Some(&child_device_id)).await?;
-            self.child_devices
-                .insert(child_device_id, c8y_internal_id.clone());
-            Ok(c8y_internal_id)
+    async fn try_get_internal_id(&mut self, device_id: String) -> Result<String, C8YRestError> {
+        let url_get_id: String = self.end_point.get_url_for_get_id(device_id);
+        if self.end_point.token.is_none() {
+            self.get_fresh_token().await?;
         }
-    }
+        let build_request = |end_point: &C8yEndPoint| -> Result<HttpRequestBuilder, C8YRestError> {
+            Ok(HttpRequestBuilder::get(&url_get_id)
+                .bearer_auth(end_point.token.clone().unwrap_or_default()))
+        };
 
-    async fn try_get_internal_id(
-        &mut self,
-        device_id: Option<&str>,
-    ) -> Result<String, C8YRestError> {
-        let url_get_id = self.end_point.get_url_for_get_id(device_id);
+        let request_builder = build_request(&self.end_point)?;
+
+        let request = request_builder.build()?;
 
-        let request_internal_id = HttpRequestBuilder::get(url_get_id);
-        let res = self.execute(request_internal_id).await?;
-        let res = res.error_for_status()?;
+        let res = match self.peers.http.await_response(request).await? {
+            Ok(response) => match response.status() {
+                StatusCode::OK => Ok(Ok(response)),
+                StatusCode::UNAUTHORIZED | StatusCode::FORBIDDEN => {
+                    self.retry_request_with_fresh_token(build_request).await
+                }
+
+                code => Err(C8YRestError::FromHttpError(
+                    tedge_http_ext::HttpError::HttpStatusError(code),
+                )),
+            },
+
+            Err(e) => Err(C8YRestError::FromHttpError(e)),
+        };
+        let res = res?.error_for_status()?;
 
         let internal_id_response: InternalIdResponse = res.json().await?;
         let internal_id = internal_id_response.id();
+
         Ok(internal_id)
     }
 
     async fn execute(
         &mut self,
-        request_builder: HttpRequestBuilder,
+        device_id: Option<String>,
+        build_request: impl Fn(&C8yEndPoint) -> Result<HttpRequestBuilder, C8YRestError>,
```

Comment:
> Yeah,its needed because get_software_list is called inside the closure and if it fails it will return an error.

There is no  `get_software_list` method nor function.

Reply:
Now that you introduced methods that build urls from an internal id and never fail, I would try what is suggested by @albinsuresh . I.e. to remove the error case by passing to the closure the internal id.

Reply:
Yes, that can be done by moving out the logic of getting the internal id into the `execute` and just sending the `internal-id` to the `closure`. However, this will make the code bit redundant. Passing the `end-point` and getting the `internal id` reduces the redundant code.